src/include/migraphx/generic_eval.hpp

```diff
@@ -0,0 +1,123 @@
+#ifndef MIGRAPHX_GUARD_MIGRAPHLIB_GENERIC_EVAL_HPP
+#define MIGRAPHX_GUARD_MIGRAPHLIB_GENERIC_EVAL_HPP
+
+#include <migraphx/module.hpp>
+#include <migraphx/stringutils.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/target.hpp>
+#include <migraphx/env.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/time.hpp>
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/pass_manager.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/register_target.hpp>
+#include <iostream>
+#include <sstream>
+#include <algorithm>
+#include <set>
+#include <utility>
+#include <unordered_set>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+template <class F>
+void print_module(const module& m, F print_func)
+{
+    std::unordered_map<instruction_ref, std::string> names;
+    int count = 0;
+
+    for(auto ins : iterator_for(m))
+    {
+        std::string var_name;
+        if(ins->name() == "@param")
+        {
+            var_name = any_cast<builtin::param>(ins->get_operator()).parameter;
+        }
+        else
+        {
+            var_name = "@" + std::to_string(count);
+            count++;
+        }
+        names.emplace(ins, var_name);
+
+        assert(std::all_of(ins->inputs().begin(),
+                           ins->inputs().end(),
+                           [&](auto arg) { return m.has_instruction(arg); }) &&
+               "PRINT_MODULE: Instruction not found");
+
+        print_func(ins, names);
+    }
+}
+
+template <class F>
+std::vector<argument> generic_eval(const module& p,
```

Comment:
These functions should be moved to the module.cpp file. The module class should have sufficient methods that the program can call into.

Reply:
What I said is incorrect here. This function should be moved into program.cpp.

Reply:
moved